python/lsst/pipe/base/connectionTypes.py

```diff
@@ -367,21 +362,23 @@ class PrerequisiteInput(BaseInput):
 
     """
 
-    lookupFunction: Optional[
-        Callable[[DatasetType, Registry, DataCoordinate, Sequence[str]], Iterable[DatasetRef]]
-    ] = None
+    lookupFunction: Callable[
+        [DatasetType, Registry, DataCoordinate, Sequence[str]], Iterable[DatasetRef]
+    ] | None = None
+
+    _connection_type_set: ClassVar[str] = "prerequisiteInputs"
 
 
 @dataclasses.dataclass(frozen=True)
 class Output(DimensionedConnection):
-    pass
+    _connection_type_set: ClassVar[str] = "outputs"
 
 
 @dataclasses.dataclass(frozen=True)
 class InitInput(BaseConnection):
-    pass
+    _connection_type_set: ClassVar[str] = "initInputs"
 
 
 @dataclasses.dataclass(frozen=True)
 class InitOutput(BaseConnection):
-    pass
+    _connection_type_set: ClassVar[str] = "initOutputs"
```

Comment:
I don't see any problem with private class attributes, but I'll concede having to use `getattr` on a string isn't great.  Would you prefer a method like
```
def _get_connections_set(self, connections: PipelineTaskConnectionsMetaclass | PipelineTaskConnections) -> set[str]:
    return connections.initOutputs
```
?

As for the duplication, I think the `BaseConnection` hierarchy and `PipelineTaskConnections` are already super tightly coupled since one is a descriptor that assumes a lot about the internals of the other, and having one attribute or method on each `BaseConnection` is much better than a series of `isinstance` checks (or the slightly prettier, but otherwise identical `match` equivalent) somewhere in `PipelineTaskConnetions`.

Reply:
Nate and I discussed this in person, so for posterity:

Nate's `type`-keyed mapping idea doesn't quite work as-is, because there are different sets on `PipelineTaskConnectionsMetaclass` and `PipelineTaskConnections`.  It could be made to work as a mapping with string values, but that brings in some of the disadvantages of the approach that's already, and we've decided to just stick with it, since it then boils down to just _slight_ philosophical/stylistic differences in preferred programming paradigms.